decidim-core/app/jobs/decidim/metric_job.rb

```diff
@@ -0,0 +1,14 @@
+# frozen_string_literal: true
+
+module Decidim
+  class MetricJob < ApplicationJob
+    queue_as :metrics
+
+    def perform(manager_class, organization_id, day = nil)
```

Comment:
As we commented on the previous review, the manager class is doing too many things.

Here's a proposal for this job:

```ruby
module Decidim
  class MetricJob < ApplicationJob
    queue_as :metrics

    def perform(metric_query_class, organization_id, day = nil)
      organization = Decidim::Organization.find(organization_id)
      query = metric_query_class.constantize.new(organization, day)

      Decidim::Metric.create!(
        day: query.day,
        metric_type: query.metric_type,
        cumulative: query.cumulative,
        quantity: query.quantity,
        organization: organization,
        participatory_space: query.participatory_space,
        related_object: query.related_object,
        category: query.category
      )
    end
  end
end
```

This way, you can set at the metrics manifest a class that will query the database and return all the needed values to save a metric. Each metric would have its own query class (you can share come logic if it's needed with a Module).

Reply:
Caution!! If this job is executed twice for same metric and day, it will duplicate metric records.

Other point is I don't see how `MetricManage` is doing *too many things*. I mean, is a base class to manage records creation, and then each metric have its own manager, for querying. Isn't this the same logic as your are saying?

Reply:
Then the metrics DB table should have a unique index on the day and metric_type columns to avoid this.

Reply:
Well, you could add an `if` clause to the job and search for an existing metric, and if it exists just return. This code was an idea, you can copy it and improve it.

About metric manage: the problem I see is that each metric doesn't need to know how to store the data, I know it's via inheritance but each class could overwrite it. Also, I don't understand the public api of the current managers (having to call `query`, `clean`, `with_context` and `registry`).

Reply:
By the way, I still think that the metric attributes could be improved:

- `cumulative` isn't this the same as loading the newest metric and return its value?
- `participatory_space` and `related_object` are redundant to me, you only need `related_object`
- `category` could go into a metadata field (like maybe participatory_space if it's needed)
- `metric_type` might be more understandable if it was named `name`
- I'd rename `quantity` to `value`

Reply:
- `cumulative` is the current total count of registries. For example, 50 users confirmed. `quantity` is the current date count regisries. For example, 3 new users confirmed today.
- `participatory_space` and `related_object` are not fully redundant. For `Proposals`, you need to know which `PS` is created. And for `Comments`, which `PS` and `Component`.
- No need for metadata fields. Remember that in next iterations, there is a category chart for proposals.
- I can change `metric_type` to `name`
- Ditto

Reply:
> cumulative is the current total count of registries. For example, 50 users confirmed. quantity is the current date count regisries. For example, 3 new users confirmed today.

Ok, I understand. What's the point of storing the `quantity` though?

Reply:
> participatory_space and related_object are not fully redundant. For Proposals, you need to know which PS is created. And for Comments, which PS and Component

Why do you need the participatory space? As an example, you could have a `organization_proposals_metric` with the number of proposals, and a `participatory_space_proposals_metric` with the count for that space. In both cases you can use the `related_object` to reference the `organization` or `participatory_space`.

Reply:
This https://github.com/decidim/decidim/issues/3398#issuecomment-393222139 shows a history chart of proposals, votes and comments created by date. so I need to know how many new registries had been created.

Reply:
> This #3398 (comment) shows a history chart of proposals, votes and comments created by date. so I need to know how many new registries had been created.

And can't you do this by just storing the value for that day (cumulative)? I'm not against storing the difference too, I just want to understand it.

Reply:
In that case I should iterate for each record for his own previous date and calculate the difference. It's easier to store it.

Reply:
But why do you need the difference to chart the data? Don't you pass a date and value for that date to the chart?

Reply:
This chart, itsquite difficult to populate without the difference, taht now is store in the `quantity` field

![chart](https://user-images.githubusercontent.com/13780790/43845062-67d211cc-9b2b-11e8-96a9-67d14a2d6301.png)

Reply:
Ohhh, I didn't see that one. OK, it makes sense now!

Reply:
I'm a bit concerned with that chart, although it's a @decidim/product issue, not a technical one: I feel like in most Decidim instances this will show a lot of zeros 😅

Reply:
That make me think @MarcReniu, if a metric is scoped to a participatory process, should we stop recording the metric at some point? The last date of the last step?

Reply:
We should not, because its still a countable registry. For example, a finished proposal is an available one (not for voting, but for viewing).

Another point of view is that only public participatory spaces are counted. So, if any time anyone is unpublish, you will see a drop down in metrics.

Reply:
But take for example a budget voting, is doesn't make a lot of sense to keep writing metrics for a process that has finished and won't have any new activity.

Reply:
Instance's issue XD. I understand what are you saying, but that means more logic related to metrics. @decidim/product must say something about this.

Reply:
I know it's more logic to metrics, and just saying "that's not my problem, let the users handle it" doesn't seem the right approach, they can't do anything to solve it.

Reply:
There is indeed a risk that many instances will end up having flat graphs due to lack of activity, but it has to serve in part to make this visible.
As for stopping recording metrics at any given time, I don't have a strong opinion on that. My first thought would be not to stop it, as long as there is some component still active, and you can you can consult the metrics of a given time range.

Reply:
@carolromero and how can we define if a component is active?

Reply:
@oriolgual I meant published, which is probably more accurate

Reply:
Hello, how is this discussing going?

Reply:
Hi, ok, correct me If I'm wrong, but I'm picturing this:
We will have, let's say a process finished, and all components remain published, but if the different actions (create new proposal, comment, endorse, etc) are disabled we won't be recording new metrics, right? @oriolgual @MarcReniu

Reply:
While the rake task is still running daily, it will keep generating records for all metrics.

Reply:
In theory we could do that @carolromero, the problem is that there's no way (I think) to list all the actions for all the components in a participatory space and know if they are active or not, that's why I suggested to stop recording metrics after the last step date (maybe we could do it for an extra month or something like this).

Reply:
If this is the case, let's do what @oriolgual proposes.